### PR TITLE
`/integrations` facelift

### PR DESCRIPTION
**Pull request by KATT**
> Sorry about the non-readable commit history 😅

- Complete rewrite of `/integrations` page
- New sandbox helper - https://calendso-git-katt-cal-549-integrations-facelift-cal.vercel.app/sandbox
- [`<List />`](https://calendso-git-katt-cal-549-integrations-facelift-cal.vercel.app/sandbox/List#main)
- [`<Input />`/`<TextField />`/`<Label />`](https://calendso-git-katt-cal-549-integrations-facelift-cal.vercel.app/sandbox/form#main) components

**Needs testing of all the integrations 🙏**

### CODE REVIEW

File: components/form/fields.tsx [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+  {
+    label: ReactNode;
+  } & React.ComponentProps<typeof Input> & {
+      labelProps?: React.ComponentProps<typeof Label>;
+    }
+>(function TextField(props, ref) {
+  const id = useId();
+  const { label, ...passThroughToInput } = props;
+
+  // TODO: use `useForm()` from RHF and get error state here too!
```
Comment (KATT) [inline components/form/fields.tsx:39]: Planning for these components to automatically handle error state from react hook form

File: lib/integrations/Apple/components/AddAppleIntegration.tsx [hunk 2]
```diff
@@ -1,11 +1,111 @@
...
+            }
+          })}>
+          <fieldset className="space-y-2" disabled={form.formState.isSubmitting}>
+            <TextField
+              required
+              type="text"
+              {...form.register("username")}
+              label="Username"
+              placeholder="rickroll"
+            />
```
Comment (KATT) [inline lib/integrations/Apple/components/AddAppleIntegration.tsx:73]: A lot nicer, right!?

File: lib/integrations/CalDav/components/AddCalDavIntegration.tsx [hunk 3]
```diff
@@ -11,8 +24,104 @@ export type AddCalDavIntegrationRequest = {
...
+              <p className="text-red-700 text-sm">
+                <span className="font-bold">Error: </span>
+                {addCalDavError.message}
+              </p>
+            )}
+            <AddCalDavIntegration
+              ref={addCalDavIntegrationRef}
+              onSubmit={handleAddCalDavIntegrationSaveButtonPress}
+            />
+          </div> */}
```
Comment (KATT) [inline lib/integrations/CalDav/components/AddCalDavIntegration.tsx:98]:
```suggestion
```

File: server/routers/viewer.tsx [hunk 4]
```diff
@@ -91,6 +94,87 @@ export const viewerRouter = createProtectedRouter()
...
+            const error = getErrorFromUnknown(_error);
+            return {
+              integration,
+              error: {
+                message: error.message,
+              },
+            };
+          }
+        })
+      );
```
Comment (KATT) [inline server/routers/viewer.tsx:160]: I'm sorry - this part is a mess - a lot [due to this](https://calendso.slack.com/archives/C022UB8FMGT/p1633687094110200)

**Review by KATT — COMMENTED** [review]
Some comments below

**Comment by PeerRich** [comment]
image

NIT: this list entry has more padding-bottom than padding-top

**Comment by PeerRich** [comment]
lets try to get this merged in and then do a design review by @Jaibles as a follow-up and do a new PR

code:
image


design:
image

**Comment by zomars** [comment]
Tried to disconnect a Stripe integration, the call succeeds but the item is still there:

image

Refreshing fixes it.

File: lib/integrations/Apple/components/AddAppleIntegration.tsx (hunk 2 above)
Comment (emrysal) [reply lib/integrations/Apple/components/AddAppleIntegration.tsx:75]: TextField default type="text" an idea :)? #evennicer

File: lib/integrations/CalDav/components/AddCalDavIntegration.tsx [hunk 5]
```diff
@@ -11,8 +24,106 @@ export type AddCalDavIntegrationRequest = {
...
+  const [errorMessage, setErrorMessage] = useState("");
+  return (
+    <Dialog {...props}>
+      <DialogContent>
+        <DialogHeader
+          title="Connect to CalDav Server"
+          subtitle="Your credentials will be stored and encrypted."
+        />
+
+        <Form
```
Comment (emrysal) [inline lib/integrations/CalDav/components/AddCalDavIntegration.tsx:44]: Idea for the onSubmit -> handleSubmit can be called automatically in Form so you can just do your handler without bothering to manually call the wrapper.

File: pages/integrations/[integration].tsx [hunk 6]
```diff
@@ -1,103 +1,9 @@
...
-                  </button>
-                </div>
-              </div>
-            </div>
-          </div>
-        </div>
-      </Shell>
-    </div>
-  );
+function RedirectPage() {
```
Comment (emrysal) [inline pages/integrations/[integration].tsx:1]: Could handle this in next.config.js rewrites 🤷?

File: lib/integrations/Apple/components/AddAppleIntegration.tsx (hunk 2 above)
Comment (KATT) [reply lib/integrations/Apple/components/AddAppleIntegration.tsx:75]: Can be done as a follow-up!

File: lib/integrations/CalDav/components/AddCalDavIntegration.tsx (hunk 5 above)
Comment (KATT) [reply lib/integrations/CalDav/components/AddCalDavIntegration.tsx:44]: I'd like that as well but I was fighting with types for too long - can do as follow-up

**Review by KATT — COMMENTED** [review]
Resolved comments from @emrysal & @zomars

**Review by emrysal — APPROVED** [review]
With the recent change to "OrNull" now approved :)

**Comment by jimi-henning** [comment]
i testet the integrations manually. all of them are working.
theres an error connecting to office365 - but that has nothing to do with this PR.

**Pull request merged**